### DISCUSSION TITLE
Logstash - send logs to external server

### DISCUSSION THREAD

**Discussion by veronikasen** [Security-Onion-Solutions/securityonion#2871] [Unsupported Versions]
I added to logstash output so it will send logs to external server. But It doesn't work and I get the following error in the logstash.log :

**[2021-02-04T19:12:55,401][WARN ][logstash.outputs.redis   ] Failed to send backlog of events to Redis {:identity=>"redis://@SecurityOnion-Master:6379/0 list:logstash:unparsed", :exception=>#<Redis::CommandError: OOM command not allowed when used memory > 'maxmemory'.>, :backtrace=>["/usr/share/logstash/vendor/bundle/jruby/2.5.0/gems/redis-4.2.2/lib/redis/client.rb:127:in `call'", "/usr/share/logstash/vendor/bundle/jruby/2.5.0/gems/redis-4.2.2/lib/redis.rb:1158:in `block in rpush'", "/usr/share/logstash/vendor/bundle/jruby/2.5.0/gems/redis-4.2.2/lib/redis.rb:69:in `block in synchronize'", "uri:classloader:/META-INF/jruby.home/lib/ruby/stdlib/monitor.rb:235:in `mon_synchronize'", "/usr/share/logstash/vendor/bundle/jruby/2.5.0/gems/redis-4.2.2/lib/redis.rb:69:in `synchronize'", "/usr/share/logstash/vendor/bundle/jruby/2.5.0/gems/redis-4.2.2/lib/redis.rb:1157:in `rpush'", "/usr/share/logstash/vendor/bundle/jruby/2.5.0/gems/logstash-output-redis-5.0.0/lib/logstash/outputs/redis.rb:152:in `flush'", "/usr/share/logstash/vendor/bundle/jruby/2.5.0/gems/stud-0.0.23/lib/stud/buffer.rb:221:in `block in buffer_flush'", "org/jruby/RubyHash.java:1415:in `each'", "/usr/share/logstash/vendor/bundle/jruby/2.5.0/gems/stud-0.0.23/lib/stud/buffer.rb:216:in `buffer_flush'", "/usr/share/logstash/vendor/bundle/jruby/2.5.0/gems/stud-0.0.23/lib/stud/buffer.rb:159:in `buffer_receive'", "/usr/share/logstash/vendor/bundle/jruby/2.5.0/gems/logstash-output-redis-5.0.0/lib/logstash/outputs/redis.rb:209:in `send_to_redis'", "/usr/share/logstash/vendor/bundle/jruby/2.5.0/gems/logstash-codec-json-3.0.5/lib/logstash/codecs/json.rb:42:in `encode'", "/usr/share/logstash/logstash-core/lib/logstash/codecs/delegator.rb:48:in `block in encode'", "org/logstash/instrument/metrics/AbstractSimpleMetricExt.java:65:in `time'", "org/logstash/instrument/metrics/AbstractNamespacedMetricExt.java:64:in `time'", "/usr/share/logstash/logstash-core/lib/logstash/codecs/delegator.rb:47:in `encode'", "/usr/share/logstash/vendor/bundle/jruby/2.5.0/gems/logstash-output-redis-5.0.0/lib/logstash/outputs/redis.rb:123:in `receive'", "/usr/share/logstash/logstash-core/lib/logstash/outputs/base.rb:105:in `block in multi_receive'", "org/jruby/RubyArray.java:1809:in `each'", "/usr/share/logstash/logstash-core/lib/logstash/outputs/base.rb:105:in `multi_receive'", "org/logstash/config/ir/compiler/OutputStrategyExt.java:138:in `multi_receive'", "org/logstash/config/ir/compiler/AbstractOutputDelegatorExt.java:121:in `multi_receive'", "/usr/share/logstash/logstash-core/lib/logstash/java_pipeline.rb:295:in `block in start_workers'"]}**

What does this error means?

**Comment by cm-ops** [comment]
Looks like Redis memory error `Redis::CommandError: OOM command not allowed when used memory > 'maxmemory'.`. Try tweaking the setting as described here https://docs.securityonion.net/en/latest/redis.html?highlight=max%20memory#tuning.

**Reply by veronikasen** [reply]
But, I added a tcp output, it shouldn't use redis. Why is it trying to use it

**Reply by weslambert** [reply]
If Logstash can't pull in logs from Redis, you won't have any logs going to any output, as they are not being processed. I'd double-check Redis, as Chris mentioned.

**Reply by veronikasen** [reply]
I gave redis 8GB, but there are still errors in the logstash log:

**[2021-02-14T10:09:39,779][WARN ][logstash.outputs.redis   ] Failed to send backlog of events to Redis {:identity=>"redis://@SecurityOnion-Master:6379/0 list:logstash:unparsed", :exception=>#<Redis::CannotConnectError: Error connecting to Redis on SecurityOnion-Master:6379 (Errno::ECONNREFUSED)>, :backtrace=>["/usr/share/logstash/vendor/bundle/jruby/2.5.0/gems/redis-4.2.2/lib/redis/client.rb:362:in `establish_connection'", "/usr/share/logstash/vendor/bundle/jruby/2.5.0/gems/redis-4.2.2/lib/redis/client.rb:107:in `block in connect'", "/usr/share/logstash/vendor/bundle/jruby/2.5.0/gems/redis-4.2.2/lib/redis/client.rb:308:in `with_reconnect'", "/usr/share/logstash/vendor/bundle/jruby/2.5.0/gems/redis-4.2.2/lib/redis/client.rb:106:in `connect'", "/usr/share/logstash/vendor/bundle/jruby/2.5.0/gems/redis-4.2.2/lib/redis/client.rb:381:in `ensure_connected'", "/usr/share/logstash/vendor/bundle/jruby/2.5.0/gems/redis-4.2.2/lib/redis/client.rb:233:in `block in process'", "/usr/share/logstash/vendor/bundle/jruby/2.5.0/gems/redis-4.2.2/lib/redis/client.rb:339:in `logging'", "/usr/share/logstash/vendor/bundle/jruby/2.5.0/gems/redis-4.2.2/lib/redis/client.rb:232:in `process'", "/usr/share/logstash/vendor/bundle/jruby/2.5.0/gems/redis-4.2.2/lib/redis/client.rb:126:in `call'", "/usr/share/logstash/vendor/bundle/jruby/2.5.0/gems/redis-4.2.2/lib/redis.rb:1125:in `block in llen'", "/usr/share/logstash/vendor/bundle/jruby/2.5.0/gems/redis-4.2.2/lib/redis.rb:69:in `block in synchronize'", "uri:classloader:/META-INF/jruby.home/lib/ruby/stdlib/monitor.rb:235:in `mon_synchronize'", "/usr/share/logstash/vendor/bundle/jruby/2.5.0/gems/redis-4.2.2/lib/redis.rb:69:in `synchronize'", "/usr/share/logstash/vendor/bundle/jruby/2.5.0/gems/redis-4.2.2/lib/redis.rb:1124:in `llen'", "/usr/share/logstash/vendor/bundle/jruby/2.5.0/gems/logstash-output-redis-5.0.0/lib/logstash/outputs/redis.rb:138:in `congestion_check'", "/usr/share/logstash/vendor/bundle/jruby/2.5.0/gems/logstash-output-redis-5.0.0/lib/logstash/outputs/redis.rb:151:in `flush'", "/usr/share/logstash/vendor/bundle/jruby/2.5.0/gems/stud-0.0.23/lib/stud/buffer.rb:221:in `block in buffer_flush'", "org/jruby/RubyHash.java:1415:in `each'", "/usr/share/logstash/vendor/bundle/jruby/2.5.0/gems/stud-0.0.23/lib/stud/buffer.rb:216:in `buffer_flush'", "/usr/share/logstash/vendor/bundle/jruby/2.5.0/gems/stud-0.0.23/lib/stud/buffer.rb:159:in `buffer_receive'", "/usr/share/logstash/vendor/bundle/jruby/2.5.0/gems/logstash-output-redis-5.0.0/lib/logstash/outputs/redis.rb:209:in `send_to_redis'", "/usr/share/logstash/vendor/bundle/jruby/2.5.0/gems/logstash-codec-json-3.0.5/lib/logstash/codecs/json.rb:42:in `encode'", "/usr/share/logstash/logstash-core/lib/logstash/codecs/delegator.rb:48:in `block in encode'", "org/logstash/instrument/metrics/AbstractSimpleMetricExt.java:65:in `time'", "org/logstash/instrument/metrics/AbstractNamespacedMetricExt.java:64:in `time'", "/usr/share/logstash/logstash-core/lib/logstash/codecs/delegator.rb:47:in `encode'", "/usr/share/logstash/vendor/bundle/jruby/2.5.0/gems/logstash-output-redis-5.0.0/lib/logstash/outputs/redis.rb:123:in `receive'", "/usr/share/logstash/logstash-core/lib/logstash/outputs/base.rb:105:in `block in multi_receive'", "org/jruby/RubyArray.java:1809:in `each'", "/usr/share/logstash/logstash-core/lib/logstash/outputs/base.rb:105:in `multi_receive'", "org/logstash/config/ir/compiler/OutputStrategyExt.java:138:in `multi_receive'", "org/logstash/config/ir/compiler/AbstractOutputDelegatorExt.java:121:in `multi_receive'", "/usr/share/logstash/logstash-core/lib/logstash/java_pipeline.rb:295:in `block in start_workers'"]}**

It happens when I add custom output conf file as seen int the screenshots, when  I comment them out everything works fine:
![image](https://user-images.githubusercontent.com/76999274/107874308-55133e00-6ec1-11eb-84fa-588e17213239.png)
![image](https://user-images.githubusercontent.com/76999274/107874358-b3402100-6ec1-11eb-9571-2794707f28e9.png)

Why does it happen?

**Reply by weslambert** [reply]
Your `if` statement should be something like:

`if [event.module] == "suricata" and [event.dataset] == "alert"`

Also, your `output` should also be filtering for `sec_onion` or some other attribute of the clones -- just because it's in the same file, might not necessarily mean it's limited to only those logs -- I believe it is still considered an output in the pipeline and is applied to all logs.

I would try modifying the logic before trying to debug further.

**Reply by veronikasen** [reply]
I filtered as you wrote, this is my new conf:
![image](https://user-images.githubusercontent.com/76999274/109640988-d7eef680-7b59-11eb-92e4-95f848ba5a60.png)

But logstash shuts down-
Logs from manager:
![image](https://user-images.githubusercontent.com/76999274/109641163-14225700-7b5a-11eb-876d-4f203ce00008.png)

Logs from a searchnode:
![image](https://user-images.githubusercontent.com/76999274/109641536-809d5600-7b5a-11eb-850b-4cb2810b4f87.png)


The logs don't give any info, which can help me.
How can I find what is the problem?
